### DISCUSSION TITLE
Testes funcionais, E2E, aceitação e caixa-preta

### DISCUSSION THREAD

**Discussion by evertonthepaula** [frontendbr/forum#1299] [Perguntas e Respostas]
Tenho pesquisado sobre testes funcionais e como implementa-los de forma decente, encontrei algumas diferenciações na nomenclatura e também algumas diferenças na explicação de cada forma de teste. Um exemplo do que aprendi até agora seria que um teste de aceitação não poderia ser automatizado pois ele não cumpre apenas requisitos funcionais descritos no projeto, abordando também necessidades reais do cliente, por isso, em teoria, deveriam ser sempre testados no cliente final, colocando a prova os requisitos definidos na documentação inicial. Entretanto testes deste tipo podem servir para descobrir casos de uso indispensáveis para o funcionamento do sistema no cliente, como já descrito, ou para cobrir apenas os requisitos da documentação, sendo que sempre o foco será atender os requisitos mínimos para o sistema ser aceito, neste caso ambos testes(funcional e aceitação) se confundem. Outra coisa que entendi a partir de um pensamento é que o teste de aceitação podem envolver por exemplo casos de problemas na usabilidade que não caberiam dentro dos testes funcionais por exemplo.

Por isto eu pergunto a vocês, qual as diferenças entres estes tipos de testes: Funcionais; E2E; Aceitação e Caixa-preta, quais suas motivações e quais efeitos colaterais de assumir ou outro.

**Comment by vNNi** [comment]
Bom, vou te falar o que eu sei e acredito estar certo.
**Teste funcional** é aquilo que tu vai fazer para validar uma funcionalidade do teu sistema, ou todas elas. Por exemplo, se o cadastramento de usuário está funcionando, fazendo todo o fluxo necessário e no final, fazer o assert.
**Teste E2E** é aquele que tu vai ir do começo ao fim do seu sistema, fazer um fluxo completo de usuários comuns e ver se não está quebrando conforme o usuário usa, é mais um teste de "front-end", pra validar telas, links etc;
**Teste de Aceitação**, como você mesmo disse, esse está mais ligado ao cliente, com as user stories, para validar se o que foi pedido e esperado, está sendo entregue.
**Teste de Caixa-preta**, é um teste de funcionalidade, praticamente o mesmo que o primeiro, onde se é verificado a execução de funcionalidades, um exemplo é ver se a funcionalidade de um campo ser obrigatório funciona ao preencher um formulário;
